### PR TITLE
emirnn.py automode=False

### PR DESCRIPTION
**Pull request by pushkalkatara**
This PR implements:

* automode=False for custom lossOp and trainOp.
* IPython notebook example to override createLossOp, createTrainOp, createCollections()
* Example contains added  Softmax L2 regularizer, AdaGrad Optimizer.

### CODE REVIEW

**Comment by pushkalkatara** [comment]
* With automode = True
auto=False

* With automode=False
auto=true

**Comment by adityakusupati** [comment]
@metastableB, please look at this PR.

File: tf/edgeml/trainer/emirnnTrainer.py [hunk 1]
```diff
@@ -68,6 +60,8 @@ def __init__(self, numTimeSteps, numOutput, graph=None,
         # Input validation
         self.supportedLosses = ['xentropy', 'l2']
         self.supportedOptimizers = ['Adam']
+
```
Comment (metastableB) [inline tf/edgeml/trainer/emirnnTrainer.py]: Could you remove these extra new lines.

File: tf/examples/EMI-RNN/00_emi_lstm_example.ipynb [hunk 2]
```diff
@@ -17,7 +17,7 @@
   },
```
Comment (metastableB) [inline tf/examples/EMI-RNN/00_emi_lstm_example.ipynb]: We do prefer retaining the outputs from training runs to give users a reference output (and reference numbers). Could you please run these notebooks afresh and push.

In the same vein, it would be nice if you push a clean run --- without cells being run twice or warnings that this notebook has already been run once before etc.

File: tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb [hunk 3]
```diff
@@ -0,0 +1,496 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Using EMI-RNN on the HAR Dataset\n",
+    "\n",
+    "This is a very simple example of how the existing EMI-RNN implementation can be used on the HAR dataset. We illustrate how to train a model that predicts on 48 step sequence in place of the 128 length baselines while attempting to predict early. For more advanced use cases which involves more sophisticated computation graphs or loss functions, please refer to the doc strings provided with the released code.\n",
```
Comment (metastableB) [inline tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb]: The last few sentences now are redundant as you've implemented `automode` and have enabled this behaviour. Since this notebook is to serve as an example to custom loss functions and optimizers, it would be nice if you could add that information here.

File: tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+    "3. `EMI_Trainer`: An instance of `EMI_Trainer` class which defines the loss functions and the training routine. This expects an `output` operator from an `EMI-RNN` implementation and attaches loss functions and training routines to it. \n",
+    "\n",
+    "To build the computation graph, we create an instance of all the above and then connect them together.\n",
+    "\n",
+    "Note that, the `EMI_BasicLSTM` class is an implementation that uses an LSTM cell and pushes the LSTM output at each step to a secondary classifier for classification. This secondary classifier is not implemented as part of `EMI_BasicLSTM` and is left to the user to define by overriding the `createExtendedGraph` method, and the `restoreExtendedgraph` method.\n",
+    "\n",
+    "Note that, to use your custom loss, overload `createLossOp` function and return the newly created operator. Use `automode=False` while creating the EMI_Trainer call.\n",
```
Comment (metastableB) [inline tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb]: This is redundant and incorrect the moment this PR is accepted. Please make this change.

File: tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    "    print(\"Round: %2d, Validation accuracy: %.4f\" % (round_, acc), end='')\n",
+    "    print(', Test Accuracy (k = %d): %f, ' % (k,  np.mean((bagPredictions == BAG_TEST).astype(int))), end='')\n",
+    "    mi_savings = (1 - NUM_TIMESTEPS / ORIGINAL_NUM_TIMESTEPS)\n",
+    "    emi_savings = getEarlySaving(predictionStep, NUM_TIMESTEPS)\n",
+    "    total_savings = mi_savings + (1 - mi_savings) * emi_savings\n",
+    "    print(\"Total Savings: %f\" % total_savings)"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (metastableB) [inline tf/examples/EMI-RNN/00_emi_lstm_example_automode_false.ipynb]: Please rerun these and do include the outputs.

**Comment by metastableB** [comment]
@pushkalkatara  Thank you for your contributions!

I have gone over your PR quickly and have some made some minor remarks. Please have a look at them before this PR can be marked as complete.

Also, I'm not particularly sure about Adam + softmax being the best configuration. I remember running Nestrov + L2 and Nestrov + softmax with some degree of success.I dont quite remember the parameter configuration there though. It would be nice to know if your observations hold for a few more hyper parameter configurations.

**Comment by pushkalkatara** [comment]
Hi, @metastableB Thanks for the review. I have made the changes as you suggested.

**Comment by pushkalkatara** [comment]
Hi, @metastableB, @adityakusupati, I am interested in contributing to the repository. If would be helpful if you could point me to unimplemented features or create Github issues. Thanks.

File: tf/examples/EMI-RNN/00_emi_lstm_example.ipynb [hunk 6]
```diff
@@ -55,7 +53,7 @@
...
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": null,
```
Comment (metastableB) [inline tf/examples/EMI-RNN/00_emi_lstm_example.ipynb]: Please include outputs to this notebook

**Comment by metastableB** [comment]
@pushkalkatara there is one more change that is required. Could you please update the PR with that. I've commented in review. I'll get back to you on this PR after testing in a few days.

@adityakusupati  could you do a quick scan ?

**Comment by pushkalkatara** [comment]
Hi, @metastableB I've added the outputs to the notebook as you suggested. Sure, thanks for the review.

**Comment by pushkalkatara** [comment]
@harsha-simhadri Thanks for the merge :D
